compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,649 @@
+                .anyMatch(d -> d.diagnosticInfo().severity().equals(DiagnosticSeverity.ERROR));
+        if (erroneousCompilation) {
+            reset();
+            return;
+        }
+
+        ModulePartNode rootNode = (ModulePartNode) ctx.node();
+        updateModulePrefix(rootNode);
+
+        for (ModuleMemberDeclarationNode member : rootNode.members()) {
+            switch (member.kind()) {
+                case FUNCTION_DEFINITION -> processFunctionDefinitionNode((FunctionDefinitionNode) member, ctx);
+                case MODULE_VAR_DECL ->
+                        processModuleVariableDeclarationNode((ModuleVariableDeclarationNode) member, ctx);
+                case TYPE_DEFINITION ->
+                        processTypeDefinitionNode((TypeDefinitionNode) member, ctx);
+            }
+        }
+        reset();
+    }
+
+    private void reset() {
+        semanticModel = null;
+        allDiagnosticInfo.clear();
+        currentLocation = null;
+        modulePrefix = Constants.CSVDATA;
+    }
+
+    private void updateModulePrefix(ModulePartNode rootNode) {
+        for (ImportDeclarationNode importDeclarationNode : rootNode.imports()) {
+            Optional<Symbol> symbol = semanticModel.symbol(importDeclarationNode);
+            symbol.filter(moduleSymbol -> moduleSymbol.kind() == SymbolKind.MODULE)
+                    .filter(moduleSymbol -> isCsvDataImport((ModuleSymbol) moduleSymbol))
+                    .ifPresent(moduleSymbol -> modulePrefix = ((ModuleSymbol) moduleSymbol).id().modulePrefix());
+        }
+    }
+
+    private void processFunctionDefinitionNode(FunctionDefinitionNode functionDefinitionNode,
+                                               SyntaxNodeAnalysisContext ctx) {
+        ChildNodeList childNodeList = functionDefinitionNode.functionBody().children();
+        for (Node node : childNodeList) {
+            if (node.kind() == SyntaxKind.LOCAL_VAR_DECL) {
+                processLocalVarDeclNode((VariableDeclarationNode) node, ctx);
+            } else if (node.kind() == SyntaxKind.ASSIGNMENT_STATEMENT) {
+                processAssignmentStmtNode((AssignmentStatementNode) node, ctx);
+            }
+        }
+    }
+
+    private void processAssignmentStmtNode(AssignmentStatementNode assignmentStatementNode,
+                                           SyntaxNodeAnalysisContext ctx) {
+        ExpressionNode expressionNode = assignmentStatementNode.expression();
+        if (!isParseFunctionOfStringSource(expressionNode)) {
+            return;
+        }
+        currentLocation = assignmentStatementNode.location();
+        Optional<Symbol> symbol = semanticModel.symbol(assignmentStatementNode.varRef());
+        if (symbol.isEmpty()) {
+            return;
+        }
+        TypeSymbol typeSymbol = ((VariableSymbol) symbol.get()).typeDescriptor();
+        validateFunctionParameterTypes(expressionNode, typeSymbol, currentLocation, ctx);
```

Comment:
Minor nit, I think you can use the monadic form of optional here
```suggestion
        semanticModel.symbol(assignmentStatementNode.varRef())
                  .map(symbol -> ((VariableSymbol)symbol).typeDescriptor())
                  .ifPresent(typeSymbol -> validateFunctionParameterTypes(expressionNode, typeSymbol, currentLocation, ctx));

```